Lib/imaplib.py

```diff
@@ -1279,6 +1385,236 @@ def print_log(self):
                 n -= 1
 
 
+class _Idler:
+    # Iterable context manager: start IDLE & produce untagged responses
+    #
+    # This iterator produces (type, datum) tuples.  They slightly differ
+    # from the tuples returned by IMAP4.response():  The second item in the
+    # tuple is a single datum, rather than a list of them, because only one
+    # untagged response is produced at a time.
+
+    def __init__(self, imap, dur=None):
+        if 'IDLE' not in imap.capabilities:
+            raise imap.error("Server does not support IDLE")
+        self._dur = dur
+        self._imap = imap
+        self._tag = None
+        self._sock_timeout = None
+        self._old_state = None
+
+    def __enter__(self):
+        imap = self._imap
+        assert not (imap._idle_responses or imap._idle_capture)
+
+        if __debug__ and imap.debug >= 4:
+            imap._mesg('idle start'
+                       + ('' if self._dur is None else f' dur={self._dur}'))
```

Comment:
Nitpick: this is kind of ugly. I think that it should be relatively clear that there's no timeout, as that can make things nicer when debugging, so let's always show it even if it's `None`. Similarly, `dur` is ambiguous, we can just make it `duration`.

```suggestion
            imap._mesg(f'idle start duration={self._dur}')
```

Reply:
I constructed the message that way because it made different code paths / behaviors easier to spot when I was watching lots of debug output during development. I suppose there's little benefit to it now, and I agree that this code snippet is a bit ugly, especially now that it uses imaplib's `if __debug__ ...` pattern instead of the logging module calls that were in my original code. 

Okay, let's change it. :)